### ISSUE TITLE
UGRID and CF data questions relevant to CMIP

### ISSUE THREAD

**Issue by taylor13** [ugrid-conventions/ugrid-conventions#20]
For CMIP we will likely encourage folks to record ugrid information for model output saved on complex grids.  I have two questions about UGRID:

1) Suppose consistent with the CF conventions I write a variable (say, surface temperature, Ts) which is carried on an unstructured grid (say , Mesh1).  CF has no mechanism for associating Ts with Mesh1.  Does the UGRID convention require definition of some attribute that directly relates Ts to Mesh1?  

2) Under UGRID, can Mesh1 and associated information be contained in an external file, or must it appear in the same file as the variable defined on that mesh?

I hope someone can provide a quick answer to this.

thanks,
Karl Taylor

**Comment by JulianKunkel** [comment]
Dear Karl,
I don't know if that's already part of the definitions of UGRID.
We have just recently been working on splitting NetCDF-4 files into e.g.
GRID + rest of data such that existing tools like ncview can seamlessly
open it without knowing that the file is split.
This splitting seems to work but we can easily share the tool with you and
discuss further the demands of the community.
Would be happy to hear your opinion.

Ideal would be that UGRID conventions provide a mean to associate those two
officially, but this is now a second step from my point of view.

Regards,
Julian

2016-01-16 22:25 GMT+01:00 taylor13 <notifications@github.com>:

> For CMIP we will likely encourage folks to record ugrid information for
> model output saved on complex grids I have two questions about UGRID:
>
> 1) Suppose consistent with the CF conventions I write a variable (say,
> surface temperature, Ts) which is carried on an unstructured grid (say ,
> Mesh1) CF has no mechanism for associating Ts with Mesh1 Does the UGRID
> convention require definition of some attribute that directly relates Ts to
> Mesh1?
>
> 2) Under UGRID, can Mesh1 and associated information be contained in an
> external file, or must it appear in the same file as the variable defined
> on that mesh?
>
> I hope someone can provide a quick answer to this
>
> thanks,
> Karl Taylor
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ugrid-conventions/ugrid-conventions/issues/20>.
>



-- 
http://wr.informatik.uni-hamburg.de/people/julian_kunkel

**Comment by rsignell-usgs** [comment]
@taylor13, In  UGRID you specify the mesh for a variable via the `mesh` attribute.   There is no way to specify that grid information is in a separate file, however. 

Here's an actual example:

http://comt.sura.org/thredds/dodsC/data/comt_1_archive/inundation_tropical/UND_ADCIRC/Hurricane_Ike_2D_final_run_without_waves.html

If you look at a particular data variable, like "zeta", you can see that the attribute "mesh" is defined, which points to the "mesh" grid variable (called "adcirc_mesh" in this example) that contains information about the mesh, like which variables hold the connectivity information and coordinates.

**Comment by taylor13** [comment]
Thanks very much Rich for the explanation and example.  We have proposed an addition to the CF-conventions defining an external_variables attribute, which would contain names of variables referenced by cell_measures attributes within the file when those variables are not actually stored in the file.  We have further proposed that external_variables could contain variable names *not* referenced by any CF-convention metadata.  
If ugrid were modified to allow the mesh variable (and all that it points to) to be stored in an external file, then for CMIP6 I would want to include the name of the mesh variable also in external_variables.  For this to make sense the ugrid convention would have to be modified to allow this.  
I would also note that the ugrid convention should probably be registered at http://www.unidata.ucar.edu/software/netcdf/conventions.html  .   Then we could include "UGRID-0.9" in the CF "Conventions" global attribute of CMIP6 files.
best regards,
Karl

**Comment by rsignell-usgs** [comment]
@taylor13, can you provide a pointer here to your `external_variables` proposal or an example of what the these files might look like?

Thanks for the tip on getting the UGRID conventions on the netcdf list.    I'll send them an e-mail and we'll see if we can get UGRID and SGRID (http://sgrid.github.io/sgrid/, 
https://github.com/sgrid/sgrid) on there as well.

**Comment by JulianKunkel** [comment]
Dear Taylor and Rich,
would it be useful for NetCDF4 files to automatically make the external
data e.g. as pointed by external_variable accessible using the
initial/original file?
This way tools accessing files have not to be aware about any conventions
and the grid can just be loaded as being part of the file?


2016-02-12 16:17 GMT+01:00 Rich Signell <notifications@github.com>:

> @taylor13 <https://github.com/taylor13>, can you provide a pointer here
> to your external_variables proposal or an example of what the these files
> might look like?
>
> Thanks for the tip on getting the UGRID conventions on the netcdf list.
> I'll send them an e-mail and we'll see if we can get UGRID and SGRID (
> http://sgrid.github.io/sgrid/,
> https://github.com/sgrid/sgrid) on there as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ugrid-conventions/ugrid-conventions/issues/20#issuecomment-183369877>
> .
>



-- 
http://wr.informatik.uni-hamburg.de/people/julian_kunkel

**Comment by taylor13** [comment]
Hi Rich and Julian,
Here's the link to the discussion of external_variables. The proposal has evolved and become much more limited over time.  Go to the very end of the discussion thread to see the "final" proposal.  http://cf-trac.llnl.gov/trac/ticket/145#comment:46

In CF we've stayed away from any standard way to refer to external links as being too fragile.  The point of the external_variables is really just to indicate that one should be able to locate somewhere the variable referred to by the attribute, but should not expect to find that variable in the file itself.

**Comment by rsignell-usgs** [comment]
@taylor13 , grrr... I couldn't access the trac site yesterday when I tried it, and it seems to be down again today
 http://www.downforeveryoneorjustme.com/http://cf-trac.llnl.gov/
![2016-02-15_8-38-42](https://cloud.githubusercontent.com/assets/1872600/13050174/b06e942c-d3bf-11e5-8ff8-09f3c365dbc4.png)
but this shouldn't be any problem adding this specification to UGRID if it's part of CF

**Comment by rsignell-usgs** [comment]
@JulianKunkel , yes, we always include the grid information as part of the dataset.   It's such a tiny amount of data compared to the massive data variables that we just stick it in every file.
